File: 02 gpio/pwm/dual_pwm.py
#!/usr/bin/env python
#-*- coding: utf-8 -*-
#If this code works, it was written by Seunghyun Lee(www.bluebaynetworks.co.kr).
#If not, I don't know who wrote it

import RPi.GPIO as GPIO 
import time 

gpwmPin = 18 
gpioPin = 13

#핀 넘버링을 BCM 방식을 사용한다.
GPIO.setmode(GPIO.BCM) 
#13, 18번 핀을 PWM용으로 설정한다. 시작 시점의 클록 속도를 384로 지정한다.
GPIO.setup(gpwmPin, GPIO.OUT)
GPIO.setup(gpioPin, GPIO.OUT)
p1 = GPIO.PWM(gpwmPin, 384)
p2 = GPIO.PWM(gpioPin, 384)

#PWM을 시작한다. 시작 시점의 duty 비는 0에서 시작한다.
p1.start(0)
p2.start(0)

angle = 0.0
while True: 
  angle += 0.1 
  #듀티비가 100%가 되면 종료한다.
  if(angle > 100.0):
    break
  #PWM의 듀비비를 조금씩 올린다.
  p1.ChangeDutyCycle(angle) 
  p2.ChangeDutyCycle(angle) 
  print "PWM Duty:", angle 
  #0.1초를 쉰다.
  time.sleep(0.1) 

#PWM을 종료한다.
p1.stop()
p2.stop()
print "PWM Test End" 
#GPIO 라이브러리를 종료한다.
GPIO.cleanup()

File: 02 gpio/GPIO-Introduction/dual_pwm.c
/**************************************
If this code works, it was written by Seunghyun Lee.
If not, I don't know who wrote it
**************************************/
#include <stdio.h>    // Used for printf() statements 
#include <stdlib.h>
#include <signal.h>
#include <wiringPi.h> // Include WiringPi library! 
#include <softPwm.h> // Include Servo library! 
 
const int gpwmPin = 18; 
const int gpioPin = 13; 

//Ctrl + C 를 누를 경우 호출되는 시그널 처리 함수 
void my_ctrl_c_handler(int sig){ // can be called asynchronously
	pwmWrite (gpwmPin, 0) ; 
	pwmWrite (gpioPin, 0) ;
	exit(0);
}

int main(void) 
{ 
	int angle = 0;
	char buf[32];

	signal(SIGINT, my_ctrl_c_handler);	//Ctrl + C Handler

	//핀 넘버링을 BCM 방식을 사용한다.
	if (wiringPiSetupGpio () < 0) 
	{ 
		printf ("Unable to setup wiringPi\n");
		return 1 ; 
	}
	//18핀을 PWM 출력 핀으로 지정한다.
	pinMode(gpwmPin, PWM_OUTPUT); 
	//PWM 모드를 PWM_MODE_MS로 지정한다.
	pwmSetMode(PWM_MODE_MS);
	//PWM 클록 속도를 384로 지정한다.
	pwmSetClock(384);
	//PWM 설정 범위를 1000으로 한다.
	pwmSetRange(1000);

	//13핀을 PWM 출력 핀으로 지정한다.
	pinMode(gpioPin, PWM_OUTPUT); 
	pwmSetMode(PWM_MODE_MS);

	//PWM 파형을 0으로 한다.(파형 발생 안됨)
	pwmWrite (gpwmPin, 0) ; 
	pwmWrite (gpioPin, 0) ; 

	//듀티비를 조금씩 증가시키는 PWM 파형을 발생시킨다. 듀티비가 100%가 되면 종료한다.
	while(1){
			angle++;
			if(angle > 1000) {
				angle = 0;
				break;
			}
			printf("%d\n", angle);
			//PWM 파형을 발생시킨다. 듀티비(%)는 angle * 100 /1000 이 된다.
			pwmWrite (gpwmPin, angle) ; 
			pwmWrite (gpioPin, angle) ;
			//10ms 쉰다.
			delay(100);
	}
	pwmWrite (gpwmPin, 0) ; 
	pwmWrite (gpioPin, 0) ;

	return 0; 
}
